src/Promise/PromiseInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+namespace GraphQL\Promise;
+
+interface PromiseInterface
+{
+    /**
+     * Waits until the promise completes if possible.
+     *
+     * @return mixed
+     * @throws \LogicException if the promise has no wait function.
+     */
+    public function wait();
+
```

Comment:
One problem I have with `wait` method is that it only makes sense for sync environment. But if we integrate complete `Promise` support it doesn't make sense as a required part of `Promise` interface (say ReactPHP promises do not have to wait).

Other problem is that we will likely need to create promises in Executor some day and then single `PromiseInterface` will be not enough. 

I had an idea to introduce `PromiseAdapter` interface instead of `PromiseInterface`, something along these lines:

```php
interface PromiseAdapter {
  public function isPromise($value);
  public function then($promise, callable $onFulfilled, callable $onRejected); // -> Promise
  public function runTillCompletion($result); // this is the same as current Executor::completePromiseIfNeeded
}
```

This gives several benefits comparing to `PromiseInterface`:
1. `runTillCompletion` can be optional depending on platform
2. We do not need `PromiseWrapper`
3. Obviously we can add other methods here like `createPromise()` or `createRejectedPromise()`, etc if we need to expend implementation (so no new stuff is required).

Concrete instance of `PromiseAdapter` will be injected to Executor.

Default NoPromiseAdapter will just throw on attempt to `then` and pass-through `$result` in `runTillCompletion`. GuzzleSyncPromiseAdapter can leverage `wait`, etc

Does it make sense? I'd like to hear your thoughs on this, maybe there are some cons which I didn't notice?